File: interfaces/cs/Socketron/SocketronData.cs
using System.Text;

namespace Socketron {
	public class SocketronData : JsonObject {
		public ushort? SequenceId {
			get { return this["sequenceId"] as ushort?; }
			set { this["sequenceId"] = value; }
		}
		public string Status {
			get { return this["status"] as string; }
			set { this["status"] = value; }
		}
		public string Type {
			get { return this["type"] as string; }
			set { this["type"] = value; }
		}
		public string Func {
			get { return this["func"] as string; }
			set { this["func"] = value; }
		}
		public object Params {
			get { return this["args"] as object; }
			set { this["args"] = value; }
		}

		public SocketronData() {
		}
		
		public static new SocketronData Parse(string json) {
			JsonObject jsonObject = JsonObject.Parse(json);
			SocketronData data = new SocketronData() {
				Status = jsonObject["status"] as string,
				Type = jsonObject["type"] as string,
				Func = jsonObject["func"] as string,
				Params = jsonObject["args"] as object
			};
			if (jsonObject["sequenceId"] != null) {
				int sequenceId = (int)jsonObject["sequenceId"];
				data.SequenceId = (ushort)sequenceId;
			}
			return data;
		}

		public new string Stringify() {
			if (this["args"] == null) {
				Remove("args");
			}
			if (this["sequenceId"] == null) {
				Remove("sequenceId");
			}
			return base.Stringify();
		}

		public LocalBuffer ToBuffer(DataType type, Encoding encoding = null) {
			if (encoding == null) {
				encoding = Encoding.UTF8;
			}
			string json = Stringify();
			int size = encoding.GetByteCount(json);
			if (size > ushort.MaxValue) {
				return null;
			}
			LocalBuffer buffer = new LocalBuffer();
			buffer.WriteUInt8((byte)type);
			buffer.WriteUInt16LE((ushort)size);
			buffer.Write(json, encoding);
			return buffer;
		}
	}
}

File: src/socketron/SocketronData.js
const Config = require("./Config");
const { DataType } = require("./Common");

class SocketronData {
	constructor(options = null) {
		this.sequenceId = null;
		this.status = null;
		this.type = "";
		this.func = "";
		this.args = null;
		this.webContents = null;

		if (options != null) {
			Object.assign(this, options);
		}
	}

	static fromJson(json) {
		data = JSON.parse(json);
		return Object.assign(new SocketronData(), data);
	}
	
	toBuffer() {
		const text = this.toJson();
		const length = Buffer.byteLength(text, Config.encoding);
		//const buffer = Buffer.from(this.toJson(), Config.encoding);
		if (length <= 0xFFFF) {
			let data = Buffer.allocUnsafe(3 + length);
			data.writeUInt8(DataType.Text16, 0);
			data.writeUInt16LE(length, 1);
			data.write(text, 3);
			return data;
		} else {
			let data = Buffer.allocUnsafe(5 + length);
			data.writeUInt8(DataType.Text32, 0);
			data.writeUInt32LE(length, 1);
			data.write(text, 5);
			return data;
		}
	}

	toJson() {
		return JSON.stringify(this, (key, value) => {
			if (value == null || value === "") {
				return undefined;
			}
			return value;
		});
	}
}

module.exports = SocketronData;
